### PR TITLE
FaultInjectionTestFS::InjectThreadSpecificReadError() should not corrupt mmaped bytes

### PR DESCRIPTION
**Pull request by siying**
Summary:
Right now FaultInjectionTestFS::InjectThreadSpecificReadError() might try to corrupt return bytes, but these bytes might be from mmapped files, which would cause segfault. Instead FaultInjectionTestFS::InjectThreadSpecificReadError() should never corrupt data unless it is in caller's buffer.

Test Plan: See db_stress still runs and make sure in a test run failurs are still injected in non-mmap cases.

### CODE REVIEW

File: utilities/fault_injection_fs.cc [hunk 1]
```diff
@@ -803,11 +803,15 @@ IOStatus FaultInjectionTestFS::InjectThreadSpecificReadError(
       *result = Slice();
       ctx->message += "inject empty result; ";
       ret_fault_injected = true;
-    } else if (!direct_io && Random::GetTLSInstance()->OneIn(7)) {
+    } else if (!direct_io && Random::GetTLSInstance()->OneIn(7) &&
+               scratch != nullptr && result->data() == scratch) {
```
Comment (ajkr) [inline utilities/fault_injection_fs.cc:807]: tangent: I hope one day mmap can use a different function that doesn't require providing an unused scratch buffer

**Review by ajkr — APPROVED** [review]
Good find! I'll check if this fixes the first two "example failures" in T101398746.

**Review by ltamasi — APPROVED** [review]

**Comment by siying** [comment]
> Good find! I'll check if this fixes the first two "example failures" in T101398746.

Kudu to @ltamasi who spent time on investigating a contbuild failure.

I am not sure how this can be related to T101398746 though.

**Comment by ajkr** [comment]
> > Good find! I'll check if this fixes the first two "example failures" in T101398746.
> 
> Kudu to @ltamasi who spent time on investigating a contbuild failure.
> 
> I am not sure how this can be related to T101398746 though.

Agreed, it seems there are several more problems to find/fix...

**Pull request closed without merging**